### PR TITLE
Sensor api example

### PR DESCRIPTION
**Pull request by zivsha**

### CODE REVIEW

File: examples/sensor-control/readme.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# Sensor Control Sample
+
+## Overview
+
+The sensor control samples aims to provide the entry point and a tutorial for using the more advanced API that the RealSense SDK provides.
+In addition to the `r2::pipeline` API, which simplifies streaming flow, the SDK provides the `rs2::context`, `rs2::device`, and `rs2::sensor` APIs that provide an abstraction for RealSense devices.
+
+In this sample we show how to use this API for querying all connected devices, their sensors, and their streams.
+Using a `rs2::sensor` one can control a camera's options, stream from it, and get the frames directly without having them synchronize (as the `rs2::pipeline` does).
```
Comment (oordan) [inline examples/sensor-control/readme.md]: without having them processed or synchronized.

File: examples/sensor-control/readme.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
+# Sensor Control Sample
+
+## Overview
+
+The sensor control samples aims to provide the entry point and a tutorial for using the more advanced API that the RealSense SDK provides.
+In addition to the `r2::pipeline` API, which simplifies streaming flow, the SDK provides the `rs2::context`, `rs2::device`, and `rs2::sensor` APIs that provide an abstraction for RealSense devices.
```
Comment (oordan) [inline examples/sensor-control/readme.md:6]: I would link here to the API overview

File: examples/sensor-control/readme.md [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+In this sample we show how to use this API for querying all connected devices, their sensors, and their streams.
+Using a `rs2::sensor` one can control a camera's options, stream from it, and get the frames directly without having them synchronize (as the `rs2::pipeline` does).
+
+## Expected Output
+
+Running this example with a connected Intel RealSense device, will output the device name and some of its details.
+For each one of the device's sensors, the output will contain a list of supported stream profiles, and if all goes well a number of frames should arrive and print their number and stream type to theo utput.
```
Comment (oordan) [inline examples/sensor-control/readme.md]: the output

File: examples/sensor-control/rs-sensor-control.cpp [hunk 4]
```diff
@@ -0,0 +1,243 @@
+// License: Apache 2.0. See LICENSE file in root directory.
+// Copyright(c) 2017 Intel Corporation. All Rights Reserved.
+
+#include <librealsense2/rs.hpp>
+#include "../example.hpp"
+#include <imgui.h>
```
Comment (oordan) [inline examples/sensor-control/rs-sensor-control.cpp]: imgui and example are required dependencies here?

File: examples/sensor-control/rs-sensor-control.cpp [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+};
+
+int main(int argc, char * argv[]) try
+{
+    // First, create a rs2::context.
+    // The context represents the current platform with respect to
+    //  connected devices
+    rs2::context ctx;
+
+    // Using the context we can get all connected devices into a device list
```
Comment (oordan) [inline examples/sensor-control/rs-sensor-control.cpp]: replace into with in

File: examples/sensor-control/rs-sensor-control.cpp [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+                    }
+                }
+            }
+
+            // We can iterate over the available profiles of a sensor
+            for (rs2::stream_profile stream_profile : sensor.get_stream_profiles())
+            {
+                // A Stream is an abstraction for a sequence of data items of a
+                //  single data type, which are ordered according to their time
+                //  of creation or arrival.
```
Comment (oordan) [inline examples/sensor-control/rs-sensor-control.cpp:139]: arrival?

File: examples/sensor-control/rs-sensor-control.cpp [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+                //    * rs2_stream::RS2_STREAM_DEPTH describes a stream of depth images
+                //    * rs2_stream::rs2_stream::RS2_STREAM_COLOR describes a stream of color images
+                //    * rs2_stream::rs2_stream::RS2_STREAM_INFRARED describes a stream of infrared images
+
+                // As mentioned, a sensor can have multiple streams.
+                // In order to distinguish between streams with the same
+                //  stream type we can use the following methods:
+
+                // 1) Each stream type can have multiple occurances.
+                //    All streams derived from a single device have distinc indices:
```
Comment (oordan) [inline examples/sensor-control/rs-sensor-control.cpp]: distinct. is it? don't color and depth have index 0?
derived? I would use "provided by"

File: examples/sensor-control/rs-sensor-control.cpp [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+                int stream_index = stream_profile.stream_index();
+
+                // 2) Each stream has a user friendly name.
+                //    The stream's name is not promised to be unique,
+                //     rather a human readable description of the stream
+                std::string stream_name = stream_profile.stream_name();
+
+                // 3) Each stream in the system, which derives from the same
```
Comment (oordan) [inline examples/sensor-control/rs-sensor-control.cpp:165]: derives from the same context? I thought that context always includes all devices. and you can instantiate multiple contexts. who gives the unique id? isn't it the singleton? 
derived is not a good word here. I think it's enough to say "each stream in the system / platform"

File: examples/sensor-control/rs-sensor-control.cpp [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+                //     rs2::context, has a unique identifier
+                //    This identifier is unique across all streams, and not
+                //     only between stream with the same stream type
```
Comment (oordan) [inline examples/sensor-control/rs-sensor-control.cpp]: this identifies is unique across all streams, regardless of the stream type.

File: examples/sensor-control/rs-sensor-control.cpp [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+                int unique_stream_id = stream_profile.unique_id(); // The unique identifier can be used for comparing two streams
+
+                std::cout << "\t\tStream #" << unique_stream_id
+                     << " is " << stream_data_type
+                     << " #" << stream_index
+                     << ", Named: \"" << stream_name << "\"" << std::endl;
+
+                // As noted, a stream is an abstraction.
+                // In order to get additional data for the specific type of
+                //  stream, a mechanism of "Is" and "As" is provided:
```
Comment (oordan) [inline examples/sensor-control/rs-sensor-control.cpp]: type of *a* stream

File: examples/sensor-control/rs-sensor-control.cpp [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+                    //  for additinal operations:
+                    std::cout << "\t\t\tThis stream is a video stream representing a stream of images with a resolution of "
+                        << video_stream_profile.width() << "x" << video_stream_profile.height() <<
+                        ", a frame rate of " << video_stream_profile.fps() << " frames per second"
+                        ", and a pixel format of: " << video_stream_profile.format() << std::endl;
+                }
+            }
+
+            // The sensor controls turning the streaming on and off
+            // To being streaming, two calls must be made with the following order:
```
Comment (oordan) [inline examples/sensor-control/rs-sensor-control.cpp]: to begin, but I would write "to start"

File: examples/sensor-control/rs-sensor-control.cpp [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+            //  1) open(stream_profiles_to_open)
+            //  2) start(function_to_handle_frames)
+
+            // In the following code we start the sensor with the first profile it provides
+            if (sensor.get_stream_profiles().size() > 0)
+            {
+                auto first_profile = *sensor.get_stream_profiles().begin();
+                // Open can be called with a single profile, or with a collection of profiles
+                // Calling open() tries to get exclusive access to the actual sensor (not only the software object of sensor)
```
Comment (oordan) [inline examples/sensor-control/rs-sensor-control.cpp]: to the physical sensor device. I would remove the (not only...).
explain here that the device may actually start running, produce data, and consume power.

File: examples/sensor-control/rs-sensor-control.cpp [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+                sensor.open(first_profile);
+
+                // In order to begin getting data from the sensor, we need to register a callback to handle frames (data)
+                // To register a callback, the sensor's start() method should be invoked.
+                // The start() method takes any type of callable object that takes a frame as its parameter
+                // NOTE:
+                //  * Since a sensor can stream multiple streams, and start
+                //     takes a single handler, multiple types of frames can arrive to the handler.
+                //  * Different streams arrive from different threads.
+                //    This behavior requires that the frame handler you provide to the start method, must be reentrant
```
Comment (oordan) [inline examples/sensor-control/rs-sensor-control.cpp]: This behavior requires the provided frame handler to the start method to be re-entrant

File: examples/sensor-control/rs-sensor-control.cpp [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+                // Calling open() tries to get exclusive access to the actual sensor (not only the software object of sensor)
+                sensor.open(first_profile);
+
+                // In order to begin getting data from the sensor, we need to register a callback to handle frames (data)
+                // To register a callback, the sensor's start() method should be invoked.
+                // The start() method takes any type of callable object that takes a frame as its parameter
+                // NOTE:
+                //  * Since a sensor can stream multiple streams, and start
+                //     takes a single handler, multiple types of frames can arrive to the handler.
+                //  * Different streams arrive from different threads.
```
Comment (oordan) [inline examples/sensor-control/rs-sensor-control.cpp]: different streams frames arrive on different threads

File: examples/sensor-control/rs-sensor-control.cpp [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+                {
+                    // We wrapped the calls to start() with try{} catch{} block for 2 reasons here:
+                    //  1) calling start() multiple times throws an exception (but we wanted to show that it compiles with different types)
+                    //  2) Start can throw an exception generally.
+
+                    std::cerr << "RealSense error calling " << e.get_failed_function() << "(" << e.get_failed_args() << "):\n    " << e.what() << std::endl;
+                }
+
+                // After calling start, frames will begin to arrive asynchronously to the callback handler.
+                // We now block the main thread from continuing, to allows frames to arrive during this time
```
Comment (oordan) [inline examples/sensor-control/rs-sensor-control.cpp]: to allow

File: examples/sensor-control/readme.md (hunk 2 above)
Comment (zivsha) [reply examples/sensor-control/readme.md:6]: which API overview?

File: examples/sensor-control/rs-sensor-control.cpp (hunk 6 above)
Comment (zivsha) [reply examples/sensor-control/rs-sensor-control.cpp:139]: I think this is originally your statement. do you want me to remove this ?

File: examples/sensor-control/rs-sensor-control.cpp (hunk 8 above)
Comment (zivsha) [reply examples/sensor-control/rs-sensor-control.cpp:165]: The context provides the unique ids, and it is not a singleton so if you have 2 contexts and you query the same device from them (I think this is supported) then you could have the same stream with 2 different ids.

File: examples/sensor-control/rs-sensor-control.cpp (hunk 10 above)
Comment (zivsha) [reply examples/sensor-control/rs-sensor-control.cpp]: Are you sure? sound right without the `a`...

Comment (oordan) [reply examples/sensor-control/rs-sensor-control.cpp]: you are right, I meant to offer "for a specific type of stream", which also not required :). ignore.

File: examples/sensor-control/rs-sensor-control.cpp (hunk 8 above)
Comment (oordan) [reply examples/sensor-control/rs-sensor-control.cpp:165]: is this a feature or a bug?

File: examples/sensor-control/rs-sensor-control.cpp (hunk 6 above)
Comment (oordan) [reply examples/sensor-control/rs-sensor-control.cpp:139]: I don't remember documenting this, but it confuses me now :)
do you think the order of frames can change between creation and arrival? it should be a bug if so, so I assume that the order is by creation. even if the TS is not...

File: examples/sensor-control/readme.md (hunk 2 above)
Comment (oordan) [reply examples/sensor-control/readme.md:6]: https://github.com/IntelRealSense/librealsense/blob/master/doc/api_arch.md
is there documentation of the sensor API design anywhere? if not I would add your comments to the api arch.

File: examples/sensor-control/rs-sensor-control.cpp (hunk 8 above)
Comment (zivsha) [reply examples/sensor-control/rs-sensor-control.cpp:165]: I believe it was by design

**Pull request closed without merging**